### ISSUE TITLE
Provide Typescript types definitions for Thing Descriptions

### ISSUE THREAD

**Issue by relu91** [w3c/wot-thing-description#1097]
Working on different Typescript-based projects I find commonly the necessity to have type definitions for Thing Descriptions, in order to a level of assurance that I am reading the TD correctly (e.g. no typos, no reading strings instead of objects, etc.). These type definitions could be reused also by Scripting APIs since there we also publish the type definitions for functions and classes but we leave out Thing Description typings. 

Other projects that could benefit from this could be:
- [node-wot](https://github.com/eclipse/thingweb.node-wot)
- [webthings/gateway](https://github.com/WebThingsIO/gateway)

And in general, the whole npm community since type definitions are also sometimes used by IDEs as hints when developing with JavaScript. 

I can volunteer to create a script that takes the JSONSchema and it convert in types, but we might need to make modifications to the schema to help a little bit the type generator.

**Comment by relu91** [comment]
Some updates about this issue. First thanks to [json-schema-to-typescript](https://www.npmjs.com/package/json-schema-to-typescript) we can obtain a good typescript definition file without substantial modifications of the current schema.  

Now that we have a working script we should think about the workflow. We can reuse the generator and add another step for generating the types but then we should also care to publish them on npm together with the right version. 
The second option is to move this issue to Scripting API. There we have already established a publication procedure for typescript packages and the TD types would only be an extension. The challenge is to keep the json schema file here in the TD repository synchronized with the types generated in Scripting API. For this purpose we can set up a Github action that automatically pulls the TD repo locally when a new change is detected in ` td-json-schema-validation.json`. 

So to recap:
- Generation in wot-thing-description:
    - Pro: no synchronization issues
    - Cons: Need to establish a proper workflow for publishing types as an npm package
- Generation in wot-scripting-api
   - Pro: publication workflow already in place
   - Cons: we need to set up a GitHub action that periodically checks the status of ` td-json-schema-validation.json

**Comment by danielpeintner** [comment]
>  * Generation in wot-scripting-api
> 
>     * Pro: publication workflow already in place
> 
>     * Cons: we need to set up a GitHub action that periodically checks the status of ` td-json-schema-validation.json`

+1

> For this purpose we can set up a Github action that automatically pulls the TD repo locally when a new change is detected in ` td-json-schema-validation.json`.

An automatic process is in fact desirable 👍

**Comment by egekorkan** [comment]
I would say that it makes more sense to automate it in the Scripting API but also link to it in the TD spec
